multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -90,35 +95,54 @@ func NewMemberClusterSetReconciler(client client.Client,
 func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	clusterSet := &mcv1alpha2.ClusterSet{}
 	err := r.Get(ctx, req.NamespacedName, clusterSet)
-	r.commonAreaLock.Lock()
-	defer r.commonAreaLock.Unlock()
+	var clusterSetNotFound bool
+	var clusterSetCreated bool
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return ctrl.Result{}, err
 		}
-		klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
-		if r.remoteCommonArea != nil {
-			if err := r.deleteMemberClusterAnnounce(ctx); err != nil {
-				// MemberClusterAnnounce could be kept in the leader cluster, if antrea-mc-controller crashes after the failure.
-				// Leader cluster will delete the stale MemberClusterAnnounce with a garbage collection mechanism in this case.
-				return ctrl.Result{}, fmt.Errorf("failed to delete MemberClusterAnnounce in the leader cluster: %v", err)
+		clusterSetNotFound = true
+	}
+
+	processClusterSet := func() error {
+		r.commonAreaLock.Lock()
+		defer r.commonAreaLock.Unlock()
+
+		if clusterSetNotFound {
+			klog.InfoS("Received ClusterSet delete", "clusterset", req.NamespacedName)
+			if err := r.cleanUpResources(ctx, clusterSetNotFound); err != nil {
+				return err
 			}
-			r.remoteCommonArea.Stop()
-			r.remoteCommonArea = nil
-			r.clusterSetConfig = nil
-			r.clusterID = common.InvalidClusterID
-			r.clusterSetID = common.InvalidClusterSetID
+			return nil
 		}
-		return ctrl.Result{}, nil
-	}
 
-	klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+		klog.InfoS("Received ClusterSet add/update", "clusterset", klog.KObj(clusterSet))
+
+		// Handle create or update
+
+		newLeader := clusterSet.Spec.Leaders[0]
+		if r.installedLeader.clusterID == newLeader.ClusterID && r.installedLeader.serverUrl == newLeader.Server &&
+			r.installedLeader.secretName == newLeader.Secret {
+			klog.V(2).InfoS("No change for leader cluster configuration")
+			return nil
+		}
+
+		if r.clusterID == common.InvalidClusterID {
+			clusterSetCreated = true
+		}
+
+		// ClusterSet deletion may fail and retry, but a ClusterSet may have been created just right before the retry.
+		// ClusterSet deletion will become an update action, so try to delete stale resources here before
+		// initilize the new ClusterSet.
+		if !clusterSetCreated {
```

Comment:
Your current code will miss the case `deleteMemberClusterAnnounce` succeeded but `cleanUpResourcesCreatedByMC` failed; and it will do cleanup even when only Secret changes. I believe we cannot just check `r.clusterID == common.InvalidClusterID`. Let me explain again.

Ideally we want:
1. `deleteMemberClusterAnnounce` when a ClusterSet is recreated but the previous `deleteMemberClusterAnnounce` failed.
2. `cleanUpResourcesCreatedByMC` when a ClusterSet is recreated before the previous `cleanUpResourcesCreatedByMC` failed.

But we should not do cleanup when only `serverUrl` or `secreteName` changes.

To achieve the above, I feel we should keep the old ClusterSet's `r.clusterID` until `cleanUpResourcesCreatedByMC` succeeds, so we can check and compare `r.clusterID` with the new ClusterID - if they do not match (inc. the case  `r.clusterID == common.InvalidClusterID`) we know a new ClusterSet is created, and then we should call  `cleanUpResources`; and if `r.clusterID != common.InvalidClusterID`, `cleanUpResourcesCreatedByMC` should be called in `cleanUpResources`.

Make sense to you?

Reply:
Yeah, refined this part. Only reset the r.clusterID when `cleanUpResourcesCreatedByMC` is done.